### PR TITLE
[CodeCamp2023-522] Support InverseForm Loss

### PR DESCRIPTION
**Pull request by CastleDream**
## Motivation

issue: [Add InverseForm Loss (Structured Boundary Aware Segmentation) #3159](https://github.com/open-mmlab/mmsegmentation/issues/3159)

discussions: [Support InverseForm Loss 支持 InverseForm Loss #522](https://github.com/open-mmlab/OpenMMLabCamp/discussions/522)

### CODE REVIEW

File: mmseg/models/losses/inverseform_loss.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+            is weighted cross-entropy loss,Default: 0.3.
+    inv_weight(float): weight for inverseform loss, Default: 0.3.
+    seg_weight(float): weight for segmentation loss which is corss-entropy
+            loss, Default: 1.
+    att_weight(float): weight for edge_attention, which were not mentioned
+            in paper, Default: 0.1.
+    loss_name (str, optional): Name of the loss item. If you want this loss
+            item to be included into the backward graph, `loss_` must be the
+            prefix of the name. Defaults to 'loss_inverseform'.
+    """
```
Comment (xiexinch) [inline mmseg/models/losses/inverseform_loss.py:220]: Might follow the indent of other docstrings.

File: mmseg/models/losses/inverseform_loss.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+
+    def bce2d(self, input, target):
+        """
+        To calculate weighted cross-entropy of two binary boundary map
+        input and target are both binary boundary map
+        Args:
+        input: pred binary boundary map
+        target: ground truth binary boundary map generated by using Sobel
+                filter and threshold on ground truth segmentation mask
+        """
```
Comment (xiexinch) [inline mmseg/models/losses/inverseform_loss.py:260]: We should add type hints. The format of the docstring is not compatible with other documents.

File: mmseg/models/losses/inverseform_loss.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+        n, c, h, w = input.size()
+
+        log_p = input.transpose(1, 2).transpose(2, 3).contiguous().view(1, -1)
+        target_t = target.transpose(1,
+                                    2).transpose(2,
+                                                 3).contiguous().view(1, -1)
```
Comment (xiexinch) [inline mmseg/models/losses/inverseform_loss.py:266]: Could use the `permute` and `reshape` functions?

File: mmseg/models/losses/inverseform_loss.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        # Perform the usual forward pass
+        x = torch.cat((x1.view(-1, 224 * 224), x2.view(-1, 224 * 224)), dim=1)
+        return x1, x2, self.fc(x)
+
+
+# Most of the code below is from the following repo:
+# https://github.com/Qualcomm-AI-research/InverseForm/blob/main/models/loss/utils.py
+
+INVERSEFORM_MODULE = os.path.join('checkpoints',
+                                  'distance_measures_regressor.pth')
```
Comment (xiexinch) [inline mmseg/models/losses/inverseform_loss.py:37]: The path should be configurable by model config files.

File: mmseg/models/losses/inverseform_loss.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+                 reduction='mean',
+                 ignore_index=255,
+                 norm=False,
+                 upper_bound=1.0,
+                 fp16=False,
+                 edge_weight=0.3,
+                 inv_weight=0.3,
+                 seg_weight=1,
+                 att_weight=0.1,
+                 loss_name='loss_inverseform'):
```
Comment (xiexinch) [inline mmseg/models/losses/inverseform_loss.py:234]: We should add type hints.

File: mmseg/models/losses/inverseform_loss.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+
+        weight[ignore_index] = 0
+
+        weight = torch.from_numpy(weight)
+        weight = weight
+        loss = F.binary_cross_entropy_with_logits(
+            log_p, target_t, weight, size_average=True)
+        return loss
+
+    def edge_attention(self, input, target, edge):
```
Comment (xiexinch) [inline mmseg/models/losses/inverseform_loss.py:296]: Might add type hints and docstring.

File: mmseg/models/losses/inverseform_loss.py [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+        n, c, h, w = input.size()
+        filler = torch.ones_like(target) * 255
+        return self.seg_loss(input,
+                             torch.where(edge.max(1)[0] > 0.8, target, filler))
```
Comment (xiexinch) [inline mmseg/models/losses/inverseform_loss.py:300]: The threshold should be configurable.

File: mmseg/models/losses/inverseform_loss.py [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+
+    def forward(self, inputs, targets, do_rmi=None):
```
Comment (xiexinch) [inline mmseg/models/losses/inverseform_loss.py:302]: Might add `**kwargs` to the parameter list, since the `BaseDecodeHead` need a unify loss interface.
https://github.com/open-mmlab/mmsegmentation/blob/main/mmseg/models/decode_heads/decode_head.py#L334

**Review by xiexinch — CHANGES_REQUESTED** [review]
Should add a unit test.

**Comment by xiexinch** [comment]
Please solve the conflict.

File: mmseg/models/losses/inverseform_loss.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        self.tiled_dim = self.resized_dim // self.tile_factor
+        self.INVERSEFORM_MODULE = inverseNet_path
+
+        inversenet_backbone = InverseNet()
+        self.inversenet = load_model_from_dict(inversenet_backbone,
+                                               self.INVERSEFORM_MODULE,
+                                               map_location)
+        for param in self.inversenet.parameters():
+            param.requires_grad = False
+
```
Comment (xiexinch) [inline mmseg/models/losses/inverseform_loss.py:111]: should add a `self._loss_name = loss_name`

File: tests/test_models/test_losses/test_inverseform_loss.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+                file.write(data)
+        progress_bar.close()
+        if total_size_in_bytes != 0 and progress_bar.n != total_size_in_bytes:
+            print('ERROR, something went wrong')
+
+    # test inverseform_loss
+    loss_class = InverseFormLoss(
+        inverseNet_path=inverseNet_path, map_location='cpu')
+    loss = loss_class(pred, target)
+    print(loss)
```
Comment (xiexinch) [inline tests/test_models/test_losses/test_inverseform_loss.py:48]: might remove `print`

File: mmseg/models/losses/inverseform_loss.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+    def forward(self, inputs: torch.tensor, targets: torch.tensor,
+                **kwargs) -> torch.Tensor:
+        """
+        Args:
+            inputs(torch.tensor): pred boundary binary map.
+            targets(torch.tensor): true boundary binary map.
+        Returns:
+            mean_square_inverse_loss(torch.tensor): InverseTransformNet loss.
+        """
+        assert inputs.shape == targets.shape
```
Comment (xiexinch) [inline mmseg/models/losses/inverseform_loss.py:121]: The shape of inputs is `(N, C, H, W)` and the shape targets is `(N, H, W)`

Comment (xiexinch) [reply mmseg/models/losses/inverseform_loss.py:121]: Should we convert the label to onehot?
https://github.com/Qualcomm-AI-research/InverseForm/blob/main/library/datasets/base_loader.py#L144
